new/on-demand-cookbook-libraries.md

```diff
@@ -0,0 +1,58 @@
+---
+RFC: unassigned
+Author: John Keiser <jkeiser@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Title
+
+On-Demand Cookbook Libraries
+
+## Motivation
+
+    As a library cookbook writer,
+    I want to have nested directories,
+    So that I can organize my code and make it more readable.
+
+    As a library cookbook writer,
+    I want to control the load order of my library,
+    So that I can have code in one file that refers to code in another file without constraining the filenames.
+
+    As a library cookbook writer,
+    I want to be able to `require` files in my library cookbook,
+    So that I don't have to deal with the pitfalls of `require_relative` (particularly that it loads things multiple times).
+
+    As a library cookbook user,
+    I want to be able to `require` files in the library cookbooks I use,
+    So that I can pick and choose what functionality I want to load.
+
+## Specification
+
+When a user creates a cookbook, they can add the following directive to the metadata:
+
+```ruby
+# metadata.rb
+name "mylibrary"
+version "0.0.1"
+eager_load_libraries false
+```
+
+If `eager_load_libraries` is false, Chef Client would instead append it to the ruby load path.  This would occur in the phase where the `libraries` files of a cookbook otherwise be automatically required in alphabetical order.  All recipes, libraries, and other Ruby code running in the Chef Client can then use `require 'filename'`` and it will load `cookbooks/mycookbook/libraries/filename` if present.
+
+If a file named `default.rb` exists in `libraries`, it will be required automatically.  No other files will be automatically loaded if `eager_load_libraries` is false.
```

Comment:
Not wild about the magic default.rb here.  I see where you want to bootstrap the library loading with a hook, but maybe this should go into the metadata parameter?

`eager_load_libraries [ 'default.rb' ]`

Allow the filenames to contain globs and then the user has control and its explicit with no magic.